mayastor/src/bdev/nexus/nexus_bdev_children.rs

```diff
@@ -165,6 +168,134 @@ impl Nexus {
+    ) -> Result<(), Error> {
+        trace!("{}: start rebuild request for {}", self.name, destination);
+
+        let source = match self
+            .children
+            .iter_mut()
+            .find(|c| c.state == ChildState::Open)
+        {
+            Some(child) => child.name.clone(),
+            None => {
+                return Err(Error::OpenChildNotFound {
+                    name: self.name.clone(),
+                })
+            }
+        };
+
+        if let Some(dst_child) =
+            self.children.iter_mut().find(|c| c.name == destination)
+        {
+            self.rebuilds.push(
+                RebuildTask::new(
+                    self.name.clone(),
+                    source,
+                    destination.to_string(),
+                    self.data_ent_offset,
+                    self.bdev.num_blocks() + self.data_ent_offset,
+                    |nexus, task| {
+                        Reactors::current().send_future(async move {
+                            Nexus::complete_rebuild(nexus, task).await;
+                        });
+                    },
+                )
+                .context(StartRebuild {
+                    child: destination.to_string(),
+                })?,
+            );
+
+            dst_child.repairing = true;
+
+            match self
+                .rebuilds
+                .iter_mut()
+                .find(|t| t.destination == destination)
+            {
+                Some(task) => {
+                    task.start();
+                    Ok(())
+                }
+                None => Err(Error::CompleteRebuild {
+                    child: destination.to_string(),
+                    reason: "rebuild task not found in the nexus".to_string(),
+                }),
+            }
+        } else {
+            Err(Error::ChildNotFound {
+                name: self.name.clone(),
+                child: destination.to_owned(),
+            })
+        }
+    }
+
+    /// On rebuild task completion it updates the child state and removes the
+    /// rebuild task in case of failure the child is left in a Faulted State
+    async fn on_rebuild_complete(&mut self, task: String) -> Result<(), Error> {
+        let task_index =
+            match self.rebuilds.iter().position(|t| t.destination == task) {
+                Some(task_index) => task_index,
+                None => {
+                    return Err(Error::CompleteRebuild {
+                        child: task,
+                        reason: "rebuild task not found in the nexus"
+                            .to_string(),
+                    });
+                }
+            };
+
+        let task = self.rebuilds.remove(task_index);
+
+        let recovered_child = match self
+            .children
+            .iter_mut()
+            .find(|c| c.name == task.destination)
+        {
+            Some(child) => child,
+            None => {
+                return Err(Error::CompleteRebuild {
```

Comment:
I wonder if this case should be an assertion. How can the destination not be present if we have completed the rebuild? Can the destination be removed during the rebuild process?

Reply:
I think it can, especially atm since there is no code to protect against it (that'll be done for CAS-126). It could be added as a test case :)